### ISSUE TITLE
Provide a way to encode system breaks and page breaks

### ISSUE THREAD

**Issue by joeberkovitz** [w3c/mnx#57]
At present CWMNX doesn't have a model for controlling pagination or system breaks. This issue serves to record the problem and serve as a nucleus for discussion.

The lack of a proposal in the spec at this time, is a reflection of the fact that page and system flow are clearly presentation properties of a score. A complicating factor is that breaks vary in placement, depending on the manner and format in which the score is rendered. Despite the uncertainty, it's clearly desirable to be able to encode some kind of information about where breaks should lie, even if it's conditional in some way.

**Comment by clnoel** [comment]
I just went and reviewed the print element in MusicXML. I think it's a good starting point, but needs work to fit what we want to do here. Here's an initial proposal:

Create a system-start element in the `<measure>` element, possibly restricted to global in some profiles. It would have the following attributes:
1. Number: system id, which corresponds across all globals or parts. Required.
2. Location: time location with respect to current measure, optional, defaults to 0 (measure start)
3. Page-start: yes/no, defaults to no. Indicates that this system starts a new page.
4. Barline: optional, indicates a barline at the start of the system.
5. Layout location: either an x-y location of the system, or a set of margins. Could possibly be content instead of attributes.

Content of this element would be staff layout instructions, including part order, distance between staves,  etc. I haven't finished thinking that through, yet, so I'm not going to be more specific.

This would represent original layout. For reflow, all of this is ignored. Any measure break is assumed to be an allowable place to do a system break (with complications for multi-metric about finding corresponding measures), and so is any mid-measure system-start element. If a measure runs longer than the desired reflow width, the reflow program will have to either compress everything, or do a mid-measure break. As mentioned in #53, we also need a way for an encoder to suggest good break points, kind of like marking the syllables in a word so that a word processor can hyphenate it properly. These would be completely optional. @cecilios disliked wrapping this into the barline element, and on reflection, so do I. Perhaps a simple `<suggested-break>` element, with a location attribute.

**Comment by joeberkovitz** [comment]
It's definitely a good time to start thinking about this more carefully. But let me ask a larger question -- is system breaking really a semantic thing? It seems to me that nearly always it's pure presentation, which we'd encode as MNX style properties rather than elements. I'd say the same about most of the data in the MusicXML `print` element: most of it only relates to a particular size page or display surface.

This doesn't mean we don't need to encode the things you're discussing -- I think we do (mostly). It's just that representing this data as semantic elements may not be the way to go.

One more messy point to consider: should an MNX document encode only breaks and system/page flow for a specific score layout (page size) or should it be able to provide more than one set of flows for different layouts in the same document?

**Comment by clnoel** [comment]
It's true that system breaks are non-semantic. I'm willing to discuss a print-layout element that will hold this info instead. And to allow more than one print-layout element, to represent different page sizes.

Another messy thing is that I've just realized that there might not be a consistent mapping of parrs to staves throughout a piece. I'm explicitly thinking of choral pieces where the voices are sometimes in unison on a single staff, are sometimes in chords on a single staff, are sometimes upstem/downstem sequences on a single staff, and are sometimes split into two (or more) separate staves. This sometimes happens within different systems of the same score. Semantically, I would consider the different voice parts to need separate part elements even when they share a staff, or even a notehead. I bring this up here because placing the parts into staves in a particular order seems to be a layout decision.

**Comment by clnoel** [comment]
I've been toying with this in my head, and I now like the `<print-layout>` option best, a completely optional element. I envision it to contain a set of `<page-layout>` elements, which contain both `<text-layout>` elements (for titles, credits, copyright info, page numbering, extra lyrics, etc.) and `<system-layout>` elements. The `<system-layout>` elements have location and width info, either absolute or relative to the previous system and the page margins. I'm not sure if it would be better to indicate starting measure number for each measure, or have child elements that indicate each contained measure and its printed width. The system-layout element will also specify which parts display in which order, but I think the definition of how that listing is done should be taken up in #97,  later.

Because this is a self-contained element it will be easy for reflow and device oriented consumers to completely skip.

**Comment by clnoel** [comment]
Hmm... just occurred to me that the draft spec states [here](https://w3c.github.io/mnx/specification/#comparisons) that all non-semantic info is moved into CSS properties. I'm not sure if we're sticking to that decision as we move forward, but I'm pretty sure a `<print-layout>` element doesn't fit into that. By that estimate, it would be better to give measures a "system-start" CSS property, except that leaves out a lot of other stuff, like multiple layouts or reflow, which wouldn't necessarily know what the "system-start" CSS property meant.

**Comment by clnoel** [comment]
After thinking about this for a while, I guess the `<print-layout>` element could function more like (or possibly use, although I hate the dependency) the jquery .addClass() function. It'll be designed to tell a consumer program to add the classes `system-start`, `indented-system-start` or other similar classes to certain measure ids (or event ids). Does that work or does it completely break the other things we're trying to do here?

**Comment by bhamblok** [comment]
@clnoel: jquery is out of the question. This spec, of course, is for other non-js languages too.

However, html, css and javascript have conquered similar issues in the past with the recent upcoming specs of "Flexbox", "CSS Grid" and, for this issues more relevant: "[Media Queries](https://www.w3.org/TR/css3-mediaqueries)".

I think we could define our spec similarly as flexbox and css-grid for the reflowing part...
But related to this issue of "hard-encoding" system breaks and pageflow, we could learn from the "Media Queries"-spec to enable us to explicitly define the values we need.

For me, it is for sure this is all non-semantic info and should be placed in CSS properties, whatever form they might take.

**Comment by joeberkovitz** [comment]
Some brief thoughts:

- @clnoel I think the approach of selectively defining style classes in different layouts can definitely  work, by allowing `<style-class>` definitions inside a layout element. We do not even need to say what the class names are; a layout `foo` can define class `fooSystemBreak` which can then be attached to any measure.  But perhaps this is an unwieldy structure... would love to hear further comments or comparisons to explicitly identifying break locations in the layout.

- @bhamblok IIRC I think @clnoel mentioned `jQuery.addClass()` to draw an analogy, not to actually use jQuery in the spec.

- I originally was optimistic about a Media Queries type approach, but have realized since then that the same score can have multiple layouts for whole-score, part A, part B, etc. Thus any media queries approach would have to be adapted to meet this use case.

- Given the digital delivery goals of MNX, is `print-` an appropriate prefix in `<print-layout>`?

**Comment by clnoel** [comment]
@joeberkovitz Sure, call it `<score-layout>` instead, then. Add a page-size element, for US Letter (8.5x11in), A4 (210x297mm), Marching Band (7x5.25in), etc. (This could be a width/height specification instead of an enumeration... choral pieces, for example, tend to be printed on all kinds of weird page sizes). Also add a pixel-width element, for scrolling page-less digital displays. You have to specify one or the other in each `<score-layout>` element. I'm not sure it's actually important for digital consumption, which will probably be doing its own reflow, but we can support it!

@bhamblok No, I wasn't actually proposing using jQuery. I'm just thinking that, in order to not break the CSS-styling that we want to be doing, we should make these `<score-layout>` metadata elements be modifiers that change the css classes of the semantic elements. That way, you can specify that a "system-start" css class on a measure will re-display the clef and key-signature before getting to its semantic content, and then specify which measures get that class added to them for this page size.

**Comment by joeberkovitz** [comment]
From the chair: we've designated this issue as a next priority to get us onto the topic of how we handle scores with multiple layouts. Page/system flow is a good forcing function for that topic.

**Comment by joeberkovitz** [comment]
I've been moving towards a unified approach to this and a variety of other issues; please let's carry the discussion to #138 (at least for now).  That proposal includes a notion of layout that is pretty similar to what @clnoel described in https://github.com/w3c/mnx/issues/57#issuecomment-380453791.

**Comment by clnoel** [comment]
Okay, since I just noticed this is in active review, I propose the following as a score-layout element, separate entirely from the notational aspects of defining the notes in each part. I made an imaginary large-orchestra score in my head that includes 8 horn parts, of which only the the Horn I part is playing for the first 10 measures. (Not a true solo, other orchestra parts are also playing then).

```
<score-layout name="Full Score" width="11in" height="17in">
    <system-layout name="Grouped Parts">
		<part-layout part="Violin 1"/>
                ...
		<group-parts type="bracket">
			<part-layout part="Horn 1"/>
			...
			<part-layout part="Horn 8"/>
		</group-parts>
		...
		<part-layout part="Piano" type="Grand"/>
	</system-layout>
	<page-layout page-number="1">
		<system system-layout-id="Grouped Parts" start-measure-id=1 end-measure-id=7 show-name="Full"/>
	</page-layout>
	<page-layout page-number="2">
		<system system-layout-id="Grouped Parts" start-measure-id=8 end-measure-id=14 show-name="Abbreviated"/>
	</page-layout>
	...
</score-layout>
<score-layout name="Condensed Score" width="8.5in" height="14in">
    <system-layout name="Unison Horns">
		<part-layout part="Violin 1"/>
               ...
		<group-parts type="chorded" name="Horn">		
			<part-layout part="Horn 1"/>
			...
			<part-layout part="Horn 8"/>
		</group-parts>
		...
		<part-layout part="Piano" type="Grand"/>
	</system-layout>
	<system-layout name="Solo Horn">
		<part-layout part="Violin 1"/>	
                ...	
		<group-parts type="bracket">
			<part-layout part="Horn 1"/>
			<group-parts type="chorded" name="Horn">		
				<part-layout part="Horn 2"/>
				...
				<part-layout part="Horn 8"/>
			</group-parts>
		</group-parts>
		...
		<part-layout part="Piano" type="Grand"/>
	</system-layout>
        <system-layout name="Split Horns">
		<part-layout part="Violin 1"/>
		<group-parts type="curly" name="Horn">
			<group-parts type="chorded" name="I II III">		
				<part-layout part="Horn 1"/>
				<part-layout part="Horn 2"/>
				<part-layout part="Horn 3"/>
			</group-parts>
			<group-parts type="split" name="IV V">		
				<part-layout part="Horn 4" force-stem="up"/>
				<part-layout part="Horn 5" force-stem="down"/>
			</group-parts>
			<group-parts type="chorded" name="VI VII VIII">		
				<part-layout part="Horn 6"/>
				<part-layout part="Horn 7"/>
				<part-layout part="Horn 8"/>
			</group-parts>
		</group-parts
		...
		<part-layout part="Piano" type="Grand"/>
	</system-layout>
	<page-layout page-number="1">
		<system system-layout-id="Solo Horn" start-measure-id=1 end-measure-id=7 show-name="Full"/>
	</page-layout>
	<page-layout page-number="2">
		<system system-layout-id="Split Horns" start-measure-id=8 end-measure-id=14 show-name="Abbreviated"/>
	</page-layout>
	...
	<page-layout page-number="12">
		<system system-layout-id="Unison Horns" start-measure-id=8 end-measure-id=14 show-name="Abbreviated"/>
	</page-layout>
</score-layout>
<score-layout name="Horn Section" width="8.5in" height="11in">
	<system-layout name="Horn Solo">
		<part-layout part="Horn 1"/>
	</system-layout>
	<system-layout name="All Horns">
		<part-layout part="Horn 1"/>
		...
		<part-layout part="Horn 8"/>
	</system-layout>
	<page-layout page-number="1">
		<system system-layout-id="Horn Solo" start-measure-id=1 end-measure-id=4 show-name="Full"/>
		<system system-layout-id="Horn Solo" start-measure-id=5 end-measure-id=8 />
		<system system-layout-id="All Horns" start-measure-id=9 end-measure-id=12 show-name="Full"/>
	</page-layout>	
	...
</score-layout>
<score-layout name="Horn 1" width="7in" height="5.5in">
	<system-layout name="Horn 1">
		<part-layout part="Horn 1" combine-full-measure-rests=true/>
	</system-layout>
	<page-layout page-number="1" show-name="Full">
		<system system-layout-id="Horn 1" start-measure-id=1 end-measure-id=4/>
		<system system-layout-id="Horn 1" start-measure-id=5 end-measure-id=8/>
		<system system-layout-id="Horn 1" start-measure-id=9 end-measure-id=12/>
		...
	</page-layout>
	...
</score-layout>
<score-layout name="Horn 2" width="7in" height="5.5in">
	<system-layout name="Horn 2">
		<part-layout part="Horn 2" combine-full-measure-rests=true/>
	</system-layout>
	<page-layout page-number="1" show-name="Full">
		<system system-layout-id="Horn 2" start-measure-id=1 end-measure-id=12/>		
		...
	</page-layout>
	...
</score-layout>
```

There's a lot here, but the critical portions are the ability to provide more than one layout for a score, to provide the ability to indicate grouped parts and nested group parts, and to combine multiple parts onto one staff, even if only temporarily.

I am not in any way tied to the names I have given the various elements and attributes. It's the _structure_ that I am proposing here. I also consider the various attributes (especially start-measure-id, end-measure-id, and show-name) to be works in progress.

If this idea meets preliminary approval, I'm willing to move forward with helping to define it further.

**Comment by bhamblok** [comment]
I'm starting to like this idea. It is flexible enough to define all imaginable layouts and it is future-proof to extend.

**Comment by clnoel** [comment]
@bhamblok  Exactly. I can also envision the need for `reduced` attributes on the `part-layouts`, and maybe an overriding staff line count attribute (especially for percussion parts) and we have been discussing in #185 the need for the `group-parts` tag to include a `barline` attribute to specify if/whether to connect a group of parts with measure barlines. I can see a lot of ways we can expand this to fit more needs as we discover them.

**Comment by clnoel** [comment]
@adrianholovaty has asked for a summary of what's been going in #185 that is relevant and specific to this issue.

The proposal that has been shaking out over there involves a `<scoreLayouts>` element that is the child of the `<mnx-common>` element (sibling to the `global` and the parts). The example below assumes that we have worked out how to define `systemLayout` elements describing which parts are visible and in which order in a given system, whether they are grouped by brackets or braces, and what the "Full" and "Abbreviated" labels are. The system layouts are also given an id for later use.


```
<scoreLayouts>
      <systemLayouts>...</systemLayouts> <!--system layouts usable in any score layout-->
      <scoreLayout name="Full Score" width="1000px" height="1700px">
            <systemLayouts>...</systemLayouts> <!--system layouts usable on any page in this layout-->
            <pages>
                  <pageLayout number="1">
		        <system startMeasureNumber="1" systemLayoutID="sysLayout1" showName="Full"/>
	          </pageLayout>
	          <pageLayout number="2">
		        <system startMeasureNumber="7" systemLayoutID="sysLayout1"  showName="Abbreviated"/>
	          </pageLayout>
	          <!--More pages-->
            </pages>  
      </scoreLayout>
      <scoreLayout name="Horns" width="8.5in" height="11in" margins=".5in,.5in,.5in,.5in">
            <systemLayouts>...</systemLayouts> <!--system layouts usable on any page in this layout-->
            <pages>
                <pageLayout number="1">
		      <system startMeasureNumber="1" systemLayoutID="sysLayout2" showName="Full"/>
                      <system startMeasureNumber="5" systemLayoutID="sysLayout3" showName="Abbreviated"/>
                      <system startMeasureNumber="9" systemLayoutID="sysLayout3" showName="Abbreviated"/>
                </pageLayout>
	        <pageLayout number="2">
		      <system startMeasureNumber="13" systemLayoutID="sysLayout3" showName="Abbreviated"/>
                       <!-More systems-->
	        </pageLayout>
	        <!--More pages-->	        	          
            </pages>  
      </scoreLayout>
      <scoreLayout name="Horn 1" width="7in" height="5.5in">
            <systemLayouts>...</systemLayouts> <!--system layouts usable on any page in this layout-->
            <pages>
                <pageLayout number="1">
		      <system startMeasureNumber="1" systemLayoutID="sysLayout25" show-name="None"/>
                      <system startMeasureNumber="6" systemLayoutID="sysLayout25" show-name="None"/>
                      <system startMeasureNumber="15" systemLayoutID="sysLayout25" show-name="None"/>
                      <!-More systems-->
                </pageLayout>
            </pages>  
      </scoreLayout>
</scoreLayouts>
```

Some notes on this from @notator 's original proposal:

* The `<page>` element could have an optional index attribute (like the Spec's `<measure>` element), but I think that's a mistake in both cases. Such indices can be deduced from the list of `<page>` (or `<measure>`) elements. I think that stipulating page numbers is a maintenance problem.
* The width/height proportion can be different for the score and parts (they have different `<scoreLayout>`s).
* If this score is printed on paper that does not have the same width/height proportion as [the layout specifies], then it will either have wider left and right margins or wider top and bottom margins.
* This `<scoreLayout>` definition does not stipulate fonts or font sizes. It simply assumes that consuming applications will be able to find settings for those that allow all the systems to have the same width, and to have heights that allow them to fit vertically on the given pages. Applications can, of course, ignore the given page breaks, and provide their own. An important application of this freedom is the writing of scores on a single page that has unlimited height (a "scroll" score). Such scores are useful for viewing on screens.

Some notes on this from @clnoel :
* We need to decide what width/height/margin units we support. I have suggested looking at SVG to see how those are specified, and @notator has supplied a link [to the SVG spec](https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/viewBox)
* We might still want to offer up a way to specify measure widths within this framework (which might be in scope for this issue).
* This doesn't address the issue of titles, credits, inter-system text blocks, and final page lyric blocks, which I think are out of scope for this issue, but could probably be worked into the `pageLayout` element when we get around to it.
* @notator proposed a `systems` collective element, to house the systems on each page, but I think this is unwise because of the other potential children of `pageLayout` that might need to be included. The presence of that is up for debate!
* I wonder if we could use a way of short-cutting the single-instrument system layout, by making it possible to specify a part directly in the `system` element instead of requiring a system layout id. On the other hand, we shouldn't build in the exceptions before we have defined the default case!

**Comment by notator** [comment]
Thanks @clnoel for the above summary!

According to the above code, system breaks are encoded by defining the number of measures in each `<system>`, and page breaks are encoded by defining the number of `<system>`s on each page.
A client application could, of course, allow its users to edit these (and all other) settings. They are only defaults that would be used when printing the score as presented in the file.
As I said:
>If [the] score is printed [to fit] on paper that does not have the same width/height proportion as [the layout specifies], then it will either have wider left and right margins or wider top and bottom margins.

So the proposal resolves this issue, and I think it could now be closed.

@clnoel I've added your list of unresolved issues to my collection of _issues to be raised later_. (I don't want to create lots of issues that would distract from the other currently active issues #188, #121 and #34.)
My unpublished issues collection now includes:
- Should my `<scores>`,  `<score>` and `<page>` elements be renamed `<scoreLayouts>` and `<scoreLayout>` and `<pageLayout>`as you suggest?
- Which width/height/margin units should be used (if any) in the `<score>` attributes? My own feeling is that absolute units are unnecessary. All that's needed is to stipulate the ideal width/height proportion. Maybe that can optionally be done by stipulating standard paper sizes (e.g. "A" would fit the score optimally on any A-series paper: A5, A4, A3, A2, A1 etc., "B" would fit the score on B-series paper, "Legal" would fit it optimally on any paper that had the width/height proportion of the "Legal" format)
- Should we be able to define measure widths? That could be interesting in connection with multi-measure rests (#121) or Forward objects. Possibly measure widths could be defined as a proportion of the system width. If the space opened by a Forward object is filled by a graphic - say a .png or a .gif -, then either the space would have to be big enough for the graphic, or the graphic would have to be shrunk to fit the space.
- Should we delete the original `<systems>` container, and make provision for "titles, credits, inter-system text blocks, and final page lyric blocks" in the `<page>` (=`<pageLayout>`) element? (I'd like to keep the `<systems>` container, and make provision for titles and credits in parallel element(s) in the `<page>` - like headers and footers. Inter-system text blocks could be defined as system-header and system-footer text blocks inside each system. Interesting idea! Thanks.)
- Can we find a way to short-cut the single-instrument system layout? (You suggest that it could be made possible to specify a part directly in the system element instead of requiring a system layout id".)

**Comment by notator** [comment]
There is now a concrete example at https://github.com/w3c/mnx/issues/185#issuecomment-780403076 showing how system breaks and page breaks can be encoded.

**Comment by adrianholovaty** [comment]
Hi all — at long last I'm ready to tackle this! Apologies for taking so long. I've read through everything in this thread and in #185. Thanks to everybody (@clnoel, @notator, @bhamblok) for the great work thinking through these thorny issues.

I'd like to propose an approach that's heavily based on the versions @clnoel and @notator came up with in various comments. Let's consider this image of music, which comes directly from the example @notator put together:

<img width="340" alt="example" src="https://user-images.githubusercontent.com/180401/128031207-f665da9f-0652-4b38-bb0d-138afff5ebc1.png">

This has a lot of nice/nasty edge cases, especially the Flute 1 and 2 parts splitting into two staves.

Here's the proposal:

```xml
<mnx>
    <global><!-- ... --></global>

    <part id="fl1" name="Flute 1" short-name="1"><!-- ... --></part>
    <part id="fl2" name="Flute 2" short-name="2"><!-- ... --></part>
    <part id="fl3" name="Flute 3" short-name="3"><!-- ... --></part>
    <part id="ob1" name="Oboe 1" short-name="1"><!-- ... --></part>
    <part id="ob2" name="Oboe 2" short-name="2"><!-- ... --></part>
    <part id="piano" name="Piano" staves="2"><!-- ... --></part>

    <system-layout id="fullscorelayout1">
        <staff-group symbol="bracket">
            <staff-group symbol="brace" name="Flutes">
                <staff parts="fl1 fl2" label="short" />
                <staff parts="fl3" label="short" />
            </staff-group>
            <staff parts="ob1 ob2" name="Oboes" label="short">
        </staff-group>
        <staff-part symbol="brace" parts="piano" label="full" />
    </system-layout>

    <system-layout id="fullscorelayout2">
        <staff-group symbol="bracket">
            <staff-group symbol="brace" name="Fl.">
                <staff parts="fl1" label="short" />
                <staff parts="fl2" label="short" />
                <staff parts="fl3" label="short" />
            </staff-group>
            <staff parts="ob1 ob2" name="Ob." label="short">
        </staff-group>
        <staff-part symbol="brace" parts="piano" label="full" />
    </system-layout>

    <system-layout id="flute1layout">
        <staff parts="fl1" label="full" />
    </system-layout>

    <score name="Full score">
        <page>
            <system measure="1" layout="fullscorelayout1" />
            <system measure="4" layout="fullscorelayout2" />
        </page>
    </score>

    <score name="Flute 1 part" layout="flute1layout">
        <page>
            <system measure="1" />
        </page>
    </score>
</mnx>
```

From the perspective of the current MNX docs, the new things here are:

* `<part>` now has `id`, `name` and `short-name` attributes. It also now has `staves`, which we hadn't yet documented!

* The `<system-layout>` element is new. This is essentially a "visual style" and has an ID so it can be referenced from a `<system>`.

* The `<score>` element is new. Each `<score>` represents a rendering of the music, with a defined subset of which `<part>`s are included, how they are visually arranged, where system breaks happen and where page breaks happen. This example has two `<score>`s, one describing the above image and another describing a theoretical Flute 1 part.

Again, this is very much based on the examples that have been designed in these GitHub threads, so I don't think there's anything revolutionary here. The main changes I made were to make it more (for lack of a better term?) "MNX-y" -- using shorter tag names that are internally consistent and removing unnecessary elements when possible.

Some scattered notes that I made while working on this:

* This example doesn't include page metadata such as margins or dimensions (A4, US Letter, etc.). I think that aspect of layout needs some more thought, particularly around which units we use (staff spaces? pixels? centimeters/inches?) and how that gets unified with other encodings of graphical spacing in MNX. But I think the structure outlined here will accommodate this metadata comfortably, so it seems we can move forward with this without needing to specify where exactly the margins/dimensions go.

* For the symbols, we can simply use the same options [as defined in MusicXML](https://w3c.github.io/musicxml/musicxml-reference/data-types/group-symbol-value/) ("bracket", "brace", etc.).

* The ability to merge multiple parts into a single staff (as done for the flute parts at the top of the image) is quite cool and would be a genuinely new feature of MNX compared to MusicXML. I'm excited about it but want to make sure that this doesn't introduce some new problems. Namely, would a rendering engine need any more information to be able to combine the two parts onto a single staff?

* For `<staff>` and `<staff-group>`, either name or label is allowed. If name is given, that's a literal string to display. If label is given, that should be either `"short"` or `"full"` and will use the corresponding name from the `<part>` (`"name"` or `"short-name"`). There's also an edge case in which both name and label are given -- see Oboes in fullscorelayout1. I'm not crazy about this solution, but I think it should be OK as long as we document "if name and label are both given, then both are displayed."

* It feels slightly weird that this supports both "two parts in a single stave" (the flutes) and "a single part on two staves" (piano) yet the mechanics of the two approaches are very different. For piano, the two-stave-ness of it is encoded in the `<part staves="2">`, while the flutes have two separate `<part>`s. This might be OK, and we should keep in mind grand-staff piano is a much more frequent use case than the flute example, but I'm just raising it as a potential inelegance/imbalance in the design.

* The `layout` attribute can go on `<score>`, `<page>` or `<system>`. This is consistent with how the `orient` and `staff` attributes generally work ([see `<event>`, for example](https://w3c.github.io/mnx/docs/mnx-reference/elements/event/)).

* In the spirit of sensible defaults, we should allow an MNX document without a `<score>` and we should allow a `<score>` without a layout. In these cases, the default rendering behavior would be to include all `<part>`s in the same order in which they appear in the markup, with line breaks and page breaks determined automatically.

* I believe this design will allow us to complete issues #188, #185, #33 and #227. Issue #34 — about encoding differences between scores and parts — would still remain open, but it feels like this design will open some possibilities in that area.

Comments and feedback? Thanks again to all of the people who came up with these ideas — I'm merely synthesizing them. :)

**Comment by clnoel** [comment]
I feel like this is a good place to restart our conversation on these topics. I don't see anything in particular other than what you have brought up in your own comments. In particular,

- I am ok with delaying the discussion of units and desired page sizes for these layouts to another issue. We can move forward without that. But we shouldn't forget about it.
- A missed part of this version of the specification was the "type" attribute put on the old `group-parts` element. That attribute would tell the renderer how to display multiple parts on the same staff (chorded or split-stem). We should probably work that back into this. Unfortunately, there is some music (I'm thinking SATB choral music in specific) that switches back and forth rather freely between those two options, so it might be more difficult than just specifying it like I did above. I'll take a bit to think about what that might look like in the new proposed format.
- I like the idea of expressing a "grand staff" (or even a triple staff) on the part itself rather than in the layout. That would make it simpler to *not* have a layout if you don't really feel like making one.
- I _really_ like the idea of being able to default layouts in the parent element to apply to all child elements. We should allow this in conjunction with overrides. For example, I think it would be great to be able to specify a Vocal-Piano system layout at the score level, and override it individually only on specific systems that do not have the vocal part. That really cuts down on the length of the XML, which is something that I remember being worried about.

--Christina
P.S. It's great to be back to this! I know the technical details of how to do the documentation needed to be addressed in order to make this a good spec, but I didn't have any stake or experience in doing one way or the other. I'm happy to be able to contribute again.

**Comment by adrianholovaty** [comment]
Christina wrote:
> A missed part of this version of the specification was the "type" attribute put on the old `group-parts` element. That attribute would tell the renderer how to display multiple parts on the same staff (chorded or split-stem). We should probably work that back into this. Unfortunately, there is some music (I'm thinking SATB choral music in specific) that switches back and forth rather freely between those two options, so it might be more difficult than just specifying it like I did above. I'll take a bit to think about what that might look like in the new proposed format.

I chatted today with @mdgood and @dspreadbury, and this came up. The "switching back and forth" indeed seems to be one of the most complicated things to encode. Do you have a rough sense of whether simplifying it to binary "chorded vs. split-stem" decision would handle 80% of common cases? 50%? Another thing to consider is the practical question of whether current notation software would even be able to handle these kinds of encodings.

> P.S. It's great to be back to this! I know the technical details of how to do the documentation needed to be addressed in order to make this a good spec, but I didn't have any stake or experience in doing one way or the other. I'm happy to be able to contribute again.

Right on — it's good to get back in action! :-D

**Comment by clnoel** [comment]
Hmm... I'm not the person to ask about a comprehensive survey of music. My _feeling_ is that most composers prefer one style or the other, and if they prefer "chorded" they would only switch to split-stem if the rhythm was different between the parts sharing the same staff, and if the rhythm is _too_ different for too long, then they switch to multiple staves. But that's just a feeling.

Parts sharing a stave should be able to handled by just about any software that could handle the *awful* `backup` musicxml element. What I'm worried about is things being able to handle parts splitting onto different staves and then rejoining. At least, that's what already gives me the most headaches on our software.

--Christina

**Comment by bhamblok** [comment]
Yes, I'm very positive about this proposal.
I have 2 minor comments:
- The `"label"`-attribute: initially I was a bit confused reading the values `"short"` or `"full"` which refer to the attributes `"name"` or "`short-name`"... Why not use `"name"` or "`short-name`" as the possible values?
- About "system breaks": I think we should make the value of the `"measure"`-attribute in a `<system>`-element a reference to an optional ID for `<measure>`-elements instead of a sequential number... that way we can easily use this workflow to define system-breaks in polymetric music.

**Comment by notator** [comment]
@adrianholovaty Thanks for https://github.com/w3c/mnx/issues/57#issuecomment-891916447.
I've also re-read this thread and #185, and am refreshing my memory...

There's a lot to discuss as we all get up to speed again. I'd like to start with this part of your code:
```
    ...
    <part id="piano" name="Piano" staves="2"><!-- ... --></part>

    <system-layout id="fullscorelayout1">
        <staff-group symbol="bracket">
            <staff-group symbol="brace" name="Flutes">
                <staff parts="fl1 fl2" label="short" />
                <staff parts="fl3" label="short" />
            </staff-group>
            <staff parts="ob1 ob2" name="Oboes" label="short">
        </staff-group>
        <staff-part symbol="brace" parts="piano" label="full" />
    </system-layout>
    ...
```
- What happens if the piano part sometimes needs to be notated on three or more staves?
- We could tackle @clnoel's "chorded vs. split-stem" problem, without complicating the single part case, if we had a separate staff type (element) for staves containing more than one part. 

I think the answer to both of these is to omit the staves attribute on the `<part id="piano" ... >` and to define the following elements (as in [my original proposal](https://github.com/w3c/mnx/issues/185#issuecomment-780403076)). Using the current MNX naming convention:
   - `<part-staff>`: a staff containing a single part
   - `<part-group-staff>`: a staff containing more than one part
   - `<part-staves>`: more than one staff for a single part (e.g. a keyboard)
   - `<staff-group>`: groups the above elements and instances of itself (this is a recursive element)
 
Parsing applications can then easily create the necessary data structures when reading the file.
(BTW: Note that the proposal for cross-staff beams in #219 uses event IDs, so would work across the whole score, not just inside `<staff-group>`, or `<part-staves>` elements. That's important if the beam crosses a system or page boundary. It would be up to the renderer to cope with such cases. )

The above example then becomes:
```
    ...
    <part id="piano" name="Piano"><!-- ... --></part>

    <system-layout id="fullscorelayout1">
        <staff-group symbol="bracket">
            <staff-group symbol="brace" name="Flutes">
                <part-group-staff parts="#fl1 #fl2" label="short" />
                <part-staff part="#fl3" label="short" />
            </staff-group>
            <part-group-staff parts="#ob1 #ob2" name="Oboes" label="short">
        </staff-group>
        <part-staves staves="2" symbol="brace" part="#piano" label="full" />
    </system-layout>
    ...

```
Notes:
1. My original example also has staff-group attributes dealing with bracket (="symbol"), barline and display types. These can all be translated directly to simple CSS attributes. Something of the sort needs to be included in the example when its finished.
2. Can "symbol" ever refer to anything other than a bracket of some sort? If not, I think its a bad name.
3. @bhamblok:
    - I agree with your misgivings about the `label` attribute, but think that my [proposal](https://github.com/notator/mnx/blob/main_ji/_ingram/mnx-examples/common1900/graphicsOnly/parts-voices-staves/partGroups.mnx) for defining `instrumentType`s for naming purposes also needs to be considered.
    - Defining "system breaks" with IDs: #155 deals with polymetric music, but I don't think it covers non-coinciding barlines (yet). Would you be happy with having the reference to an optional <measure> ID be an _alternative_ to the sequential number? That would simplify things a lot for renderers that always align their barlines. There's always a lot of overhead managing IDs... Maybe there's a better way to do this -- perhaps by defining lists of parallel measures?).
4. I still feel strongly that references to IDs should be preceded by a '#' character. This a generally accepted convention, so disregarding it will break most parsing libraries. The convention also makes the XML easier to read for humans -- one can directly see an ID for what it is. If we don't start now, the amount of work needed to correct the docs later is only going to increase.

Enough for now!
All the best. :-)
James

**Comment by notator** [comment]
I'm wrong about IDs having to be preceded by a '#' character.
See [https://stackoverflow.com/questions/22091450/xml-referencing-to-an-id-with-or-not](https://stackoverflow.com/questions/22091450/xml-referencing-to-an-id-with-or-not)

As I now understand it, the XML Schema attribute types `xs:ID`, `xs:IDREF` and `xs:IDREFS` should be used (in the schema)  for validation purposes.

**Comment by ahankinson** [comment]
See my comment here: https://github.com/w3c/mnx/pull/190#issuecomment-748117620

Hash fragments are not part of the XML spec, but they do allow `xml:id` to tie in to the URI spec. Depending on your use cases, this may or may not be desirable. As I mentioned, members of the [MEI community](https://dl.acm.org/doi/10.1145/3273024.3273038) is using this functionality to tie music notation encoding into the Linked Data world, allowing each musical element to have a unique URI associated with them based on their `xml:id` value, and then used to build graph-based representations of music notation. 

This convention is also used as part of the [XPointer](https://www.oreilly.com/library/view/learning-xml/0596000464/0596000464_learnxml-ID-3.3.html) spec, as mentioned in the StackOverflow answer

If all you are doing is making internal references to local files, though, and don't want to create IDs that also function as URIs, you don't need them.

**Comment by adrianholovaty** [comment]
Hi all — in pull request #245, I've updated the MNX documentation to add all of the elements/attributes per this discussion:

* `<system-layout>`
* `<staff-group>`
* `<staff>`
* `<score>`
* `<page>`
* `<system>`

You can dive into the docs (as they exist in the pull request) via this link:
https://cdn.githubraw.com/w3c/mnx/system-layouts/docs/mnx-reference/examples/system-layouts/
